### ISSUE TITLE
Group inheritance and Indirect group membership

### ISSUE THREAD

**Issue by nitnelave** [nitnelave/lldap#13]
More of a nice-to-have, I don't expect many users to have a very complex group membership structure that requires that.

We can make a group inherit from another one, or be a subgroup of another one: if group A has subgroup B, then users in B are indirectly part of group A.

To avoid making too many requests for reading (common case), we can keep a "resloved" membership table that contains both direct and indirect memberships. This can be updated when adding a user to a group, and can be reset when deleting a user or a group.

Another way to do it is to have just the resolved group inheritance, not the full user one; then a user is part of group A if:
 - They are explicitly a member of A.
 - They are explicitly a member of a group that resolves to inherit from A.
That's achievable with a single query, with a join.
This would only require updates/rebuilding when adding/deleting groups/group inheritance.

**Comment by nitnelave** [comment]
https://ldapwiki.com/wiki/LDAP_MATCHING_RULE_IN_CHAIN

**Comment by nitnelave** [comment]
Design for the feature, feel free to comment on it: https://docs.google.com/document/d/1xkUFGdjfbTI5rC_sh1QIRET6njsK3Ci2ueGAWO-FJY4/edit?usp=drivesdk

**Comment by lordratner** [comment]
1000% want this feature. It's really the only thing I think is truly "missing" from LLDAP.

**Comment by jacobw** [comment]
Link has changed:
https://ldapwiki.com/wiki/Wiki.jsp?page=LDAP_MATCHING_RULE_IN_CHAIN

**Comment by lordratner** [comment]
Is this on the radar still? Just wondering for planning purposes.

**Comment by nitnelave** [comment]
Yes, it is. However, expect development to be slow in the next few months (like it has been in the past few months) due to personal circumstances.

**Comment by lordratner** [comment]
> Yes, it is. However, expect development to be slow in the next few months (like it has been in the past few months) due to personal circumstances.

No worries. LLDAP is the best game in town, thank you for what you've done already.

**Comment by lordratner** [comment]
Ew, I looked this up and had no idea it has already been a year! Are we getting old?

Loving the changes in 0.6. LLDAP is **still** the best game in town. Wondering if nested groups are planned for the next release. 

Cheers!

**Comment by nitnelave** [comment]
I don't have that much dev time, but they're on my to-do list once I do (when the kid grows up a bit more, you can't rush these things!)

**Comment by sailorbob134280** [comment]
Looking forward to it, and thanks for the hard work on LLDAP! This is the only feature I feel like I miss right now.

**Comment by jsmith212** [comment]
Just looking to follow up on this one. Any updates? Is this still planned? This makes managing a lot of services really simple when adding new members since I would be able to just assign them a catch-all role (Reader, Developer, SRE, Admin) and then manage fine-grained group memberships under the coarse organizational roles.

**Comment by nitnelave** [comment]
This is still planned, but it is a complex change and we're starved for dev time.

**Comment by jsmith212** [comment]
Wanted to ping, any way to get this looked at? Just deployed another instance for a small company and I ran across this one again. I understand the dev time limitations and it's not a blocker for us yet, just wanted to check in and see if it's on the roadmap any time soon.

**Comment by nitnelave** [comment]
It is on the roadmap, it's just not so much a roadmap than a whisper of distant dreams... The car has broken down by the deserted highway, and I wait, thumb outstretched, for a stranger to reach out and help. 

Lyrical tendencies aside, LLDAP is not high enough in my priority list to make the cut very often, so dev work has stalled. I'd like to do more, but these days my free time is a carefully provisioned portion to treasure, when I have any.

**Comment by lordratner** [comment]
> It is on the roadmap, it's just not so much a roadmap than a whisper of distant dreams... The car has broken down by the deserted highway, and I wait, thumb outstretched, for a stranger to reach out and help.
> 
> Lyrical tendencies aside, LLDAP is not high enough in my priority list to make the cut very often, so dev work has stalled. I'd like to do more, but these days my free time is a carefully provisioned portion to treasure, when I have any.

Would you be averse to having it "vibe coded?"

You'd still have to review the code either way, but I could hammer it out in a few days (once I started). It's something I already do a lot for my other projects (all airline pilot related) and I'm pretty good at keeping the programming tight and focused due to my previous experience.

I'm not going to spend the time if you wouldn't want it in the repo, so just let me know.

**Comment by nitnelave** [comment]
I'm not against AI generated code, but I'm against thoughtless code; I expect high quality PRs and I'll review them the same however you wrote it. Given that I have to maintain the code after that, I have a relatively high bar in terms of code; though AI can be helpful there, and I will of course do my best to help you get there.

I'll do my best to review it soon, but I can't promise anything. Though now may be a good time since I have some holidays coming up.

Make sure to read https://github.com/lldap/lldap/issues/13#issuecomment-1200435994 and comment on it if necessary.
